### ISSUE TITLE
Cannot connect to advertisement in BLE manager (Mi Flora App)

### ISSUE THREAD

**Issue by koktaildotcom** [athombv/apps-sdk-issues#117]
With the latest firmware (v4.2.0-rc5+) i cannot connect to the advertisement anymore. It gives a timeout.

```
            const advertisement = await Homey.ManagerBLE.find(device.getAddress()).then(function (advertisement) {
                return advertisement;
            })

            const peripheral = await advertisement.connect();
```

Expect behaviour:
I get the peripheral from the advertisement.

Current behaviour
I got a BLE Timeout:

[Error: BLETimeout
   at Remote IPC Process
] {
  message: 'BLETimeout'
}

**Comment by koktaildotcom** [comment]
Looks link the `peripheral.disconnect()` doesn't work and that the seconds time the connection is made the peripheral is still connected and that `advertisement.connect()` result in a timeout.

@WeeJeWel can you verify that?

**Comment by koktaildotcom** [comment]
Ignore that. 
After a reboot of homey the sensors are connected correctly the first time. Looks like Homey keep some reference to the device.

**Comment by koktaildotcom** [comment]
I've made a custom branch for tracking the issue. You can find it here: https://github.com/koktaildotcom/homey-mi-flora/tree/feature/ble-timeout-issue.

Steps to reproduce:
1. Restart homey without any `xiaomi mi flora` app installed.
1. Wait until homey is booted properly.
1. Run the App.
1. Install one xiaomi mi flora sensor. (the first update sequence starts)
1. Wait until the second update sequence. (in this branch its a 5 second delay)
1. The error occur by all the devices (it can take some while because its a timeout)

If this doesn't work some how try rebooting homey with the app and sensor installed.

This produces a log like:

```
─────────────── Logging stdout & stderr ───────────────
Successfully init HomeyMiFlora version: 2.1.13
[extra comment for testing] ---> This is update sequence no:1.
 
-----------------------------------------------------------------
| New update sequence 
-----------------------------------------------------------------
reduce
#########################################
# update device: Mi Flora Sensor
# firmware: 3.2.1
#########################################
call handleUpdateSequence
handleUpdateSequence
find
distance = 6.641101633089944 meter
connecting to advertisement
dataService
realtime
data
DATA_CHARACTERISTIC_UUID::read
{
  measure_temperature: 22.3,
  measure_luminance: 194,
  flora_measure_fertility: 195,
  flora_measure_moisture: 9
}
FIRMWARE_CHARACTERISTIC_UUID::read
{
  firmware_version: '3.2.1',
  last_updated: '2020-05-15T18:08:29.102Z',
  uuid: 'c47c8d661988',
  battery: 96
}
call disconnectPeripheral
try to disconnect peripheral
disconnect peripheral
[extra comment for testing] ---> Looks like it happen after the disconnect
Device sync complete in: 1.555 seconds
All devices are synced complete in: 1.64 seconds
------------------------------------------------------------------------------------------------------------------------------------------------
set timeout 2
[extra comment for testing] ---> Set the update interval to 5 seconds for testing purposes.
[extra comment for testing] ---> This is update sequence no:2.
 
-----------------------------------------------------------------
| New update sequence 
-----------------------------------------------------------------
reduce
#########################################
# update device: Mi Flora Sensor
# firmware: 3.2.1
#########################################
call handleUpdateSequence
handleUpdateSequence
find
distance = 6.641101633089944 meter
connecting to advertisement
dataService
realtime
data
DATA_CHARACTERISTIC_UUID::read
{
  measure_temperature: 22.2,
  measure_luminance: 175,
  flora_measure_fertility: 196,
  flora_measure_moisture: 9
}
FIRMWARE_CHARACTERISTIC_UUID::read
{
  firmware_version: '3.2.1',
  last_updated: '2020-05-15T18:08:38.830Z',
  uuid: 'c47c8d661988',
  battery: 96
}
call disconnectPeripheral
try to disconnect peripheral
disconnect peripheral
[extra comment for testing] ---> Looks like it happen after the disconnect
Device sync complete in: 6.59 seconds
All devices are synced complete in: 6.672 seconds
------------------------------------------------------------------------------------------------------------------------------------------------
set timeout 2
[extra comment for testing] ---> Set the update interval to 5 seconds for testing purposes.
[extra comment for testing] ---> This is update sequence no:3.
 
-----------------------------------------------------------------
| New update sequence 
-----------------------------------------------------------------
reduce
#########################################
# update device: Mi Flora Sensor
# firmware: 3.2.1
#########################################
call handleUpdateSequence
handleUpdateSequence
find
distance = 6.641101633089944 meter
connecting to advertisement
try to disconnect peripheral
disconnect peripheral
[extra comment for testing] ---> Looks like it happen after the disconnect
timeout, retry again 1
[Error: Peripheral is not connected (state is disconnected)
   at Remote IPC Process
] {
  message: 'Peripheral is not connected (state is disconnected)'
}
```

**Comment by koktaildotcom** [comment]
**At the last update sequence there is an error:**
[Error: Peripheral is not connected (state is disconnected)
   at Remote IPC Process
] {
  message: 'Peripheral is not connected (state is disconnected)'
}

**Comment by koktaildotcom** [comment]
This is still an issue. Have you test it on an Homey 2016? Looks like all complains are 2019 or Pro's.

**Comment by robiebab** [comment]
I have a homey from 2016 and have sometimes a timeout, if i restart the app it gives a message timeout 30000. The only way foe me to fix it, is to reboot homey. Then it works for a few min or a few days. If you need more information, just ask :)

**Comment by shaarkys** [comment]
`

nodeVersion | "v12.16.1"
-- | --
homeyVersion | "4.2.0"
homeyModelId | "homey3s"
homeyModelName | "Homey (Early 2019)"

`

```
#########################################
# update device: Mi Flora Sensor - Strawberries
# firmware: 3.2.2
#########################################
call handleUpdateSequence
handleUpdateSequence
find
distance = 9.334850610218087 meter
connect
disconnectPeripheral not registered yet
timeout, retry again 3
[Error: BLETimeout
   at Remote IPC Process
] {
  message: 'BLETimeout'
}
Error: Error: Error: Max retries (3) exceeded, no success
    at /app.js:251:35
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async HomeyMiFlora.updateDevice (/app.js:237:16)
    at async HomeyMiFlora.updateDevices (/app.js:205:16)
reduce
```

This happens right after **connect** attempt : 
`const advertisement = await Homey.ManagerBLE.find(device.getAddress(), 10000).then(function (advertisement) {
                return advertisement;
            });
const peripheral = await advertisement.connect(); `

This somehow started with 4.2 version and get worse after each RC ... I think it stopped completely like month ago or so (RC12 maybe?)

**Comment by shaarkys** [comment]
**With the test build (tested not after reboot however)** 

```
─────────────── Logging stdout & stderr ───────────────
Successfully init HomeyMiFlora version: 2.1.13
No devices found yet.
set timeout 4
not running
No devices found yet.
set timeout 4
registeredDevice Mi Flora Sensor - Tomato
registeredDevice Mi Flora Sensor - Strawberries
registeredDevice Mi Flora Sensor - Grass
not running
[extra comment for testing] ---> This is update sequence no:1.




-----------------------------------------------------------------
| New update sequence
-----------------------------------------------------------------
reduce
#########################################
# update device: Mi Flora Sensor - Tomato
# firmware: 3.2.2
#########################################
call handleUpdateSequence
handleUpdateSequence
find
estimatedDistance = 10.1 meter
connecting to advertisement
disconnectPeripheral not registered yet
timeout, retry again 1
[Error: BLETimeout
   at Remote IPC Process
] {
  message: 'BLETimeout'
}
#########################################
# update device: Mi Flora Sensor - Tomato
# firmware: 3.2.2
#########################################
call handleUpdateSequence
handleUpdateSequence
find
estimatedDistance = 10.1 meter
connecting to advertisement
....
```

**Comment by koktaildotcom** [comment]
@WeeJeWel would you investigate this issue with additional informatie please? Can you try to keep the app running for some time. Looks like it can work for some time. Can it be related to the node update? 

Any change i can test something?

**Comment by shaarkys** [comment]
Just if that would help - here is how it behaves after daily reboot - for some time it reads values and then BLETimeout occurs.

![obrazek](https://user-images.githubusercontent.com/14859895/84378228-19163380-abe4-11ea-811e-63344a8593fe.png)

**Comment by koktaildotcom** [comment]
The community did some testing. If you disable all flows (“when value is below set value”) it looks like the BLE timeout isn't kicking in that quick. However with some tests after 21 updates with an interval of 15 minutes the BLE timeout kicks in again.

**Comment by koktaildotcom** [comment]
When do this issue get some love? I've got a lot of complains here. I cannot do anything about it.

**Comment by homeyjoenka** [comment]
Hi 

every day homey loses contact with the flower, we restart homey en dan connect he for couple of hours

**Comment by akrug42** [comment]
Hi
Any Progress on this issue. I have the same issue.  After couple updates from the sensor the update process seams to be stuck.
Is there anything I can do to help speeding up the bug fixing process?

**Comment by davidhalford** [comment]
Just installed the app and have 1 unit (not the pot, the other thing). Same issue; initial connect works, nothing after that. If you need any info from my side, please let me know, as I'd like to keep my plant alive :)

**Comment by RobvandeGoor** [comment]
I am using a Homey Pro and not able to read the sensor data anymore. Reboot Homey several times and also the app. 1 sensor is working sometimes and from the 2nd sensor I do not receive any data. It is also not possible to add new sensors (received today) 😔

**Comment by homeyjoenka** [comment]
We have to same problem its not fine

**Comment by Raycokes2000** [comment]
Same here too, sensor status looks frozen even after several reboots of homey early19 pro, Flora app and even battery change of sensors do not work

**Comment by niekvugteveen** [comment]
Same here. After restart I get one reading, then nothing. This is taking too long. It makes devices unusable.

**Comment by fmvdb** [comment]
Just received 3 Mi Flora Sensors today and ran into this issue. Data is update in Homey one time. On the smartphone the data is read flawless with the app from Mi (Flower Care).

Looking forward to a fix!

**Comment by tilcoating** [comment]
Three sensors, once updated upon install... Please solve, Athom. Bluetooth is quite useless in this way.

**Comment by dbuekenh** [comment]
Same problem here with 2 sensors on a homey with latest release

😕

**Comment by koktaildotcom** [comment]
@WeeJeWel is there any progress on this issue? It's been there for half a year. I still get a lot of complains about it.

**Comment by shaarkys** [comment]
Btw, I can confirm the same happening as well on v5.0.0-rc.34 (which is understandable as Bluetooth stack hasn't been touched), yet just for the record.

**Comment by geurti64** [comment]
![F597670B-D907-4406-B3F6-A8CAF8A5A4AF](https://user-images.githubusercontent.com/44948016/97078678-0548f800-15ee-11eb-9999-4bace8d9a5d5.jpeg)
My Homey 2018 got every week a reboot to make sure I get the plant updated again. As you can see in the screenshot not all the updates are 6 days old. Some are much older and one last update is even two months old.
I wonder if it's only a BLE problem. But I've not enough knowledge about it to say why. What surprises me is that out of four I get two updates. Why are the other two updates so much older?

**Comment by flits** [comment]
Worked without any problem. But looks like it stopped working 4 days ago. Status is not being updated. Only one synch after a Homey reboot and then it stops updating.

**Comment by koktaildotcom** [comment]
In the Homey v5.0.0-rc.41 release: 

> [Core] Fixed an issue with managing BLE connections

I've tested it on the SDK3 and i don't get timeouts anymore. Many thanks for solving this issue. Can you give us some insights what happened?

**Comment by shaarkys** [comment]
For me it is still the same, I'm on v5 38 since release... :-( After reboot it reads values and then nothing.

**Comment by geurti64** [comment]
> For me it is still the same, I'm on v5 38 since release... :-( After reboot it reads values and then nothing.

So still no reason for me to get v5 🤓

**Comment by koktaildotcom** [comment]
I misinformed you, my development Homey is on v5.0.0-rc41 and is not stable for experimental release yet. We have to wait a little longer.

**Comment by shaarkys** [comment]
No problem, it's great they fixed it, no matter which RC :-) Thank you for sharing.

**Comment by mrlukasbos** [comment]
The BLE connection issue should be resolved in V5.0.0-rc.41 and will be shipped with the next beta release. Feel free to reopen this thread if any problems still occur. 

Thank you very much for reporting the issue and for providing the detailed information.
